### PR TITLE
Biomolecules.

### PR DESCRIPTION
**Pull request by stuartjamesmcnicholas**

### CODE REVIEW

File: baby-gru/src/components/card/MoorhenMoleculeRepresentationSettingsCard.tsx [hunk 1]
```diff
@@ -69,6 +70,14 @@ export const MoorhenMoleculeRepresentationSettingsCard = (props: {
         }
     }, [symmetryOn])
 
+    useEffect(() => {
+        if (props.molecule.biomolOn !== biomolOn) {
+            if(biomolOn)
+                props.molecule.symmetryOn = false
```
Comment (FilomenoSanchez) [inline baby-gru/src/components/card/MoorhenMoleculeRepresentationSettingsCard.tsx:76]: It seems that `molecule.symmetryOn = false` is already happening inside `molecule.toggleBiomolcule` in which case this here is unnecessary?

File: baby-gru/src/utils/MoorhenMolecule.ts [hunk 2]
```diff
@@ -226,11 +275,23 @@ export class MoorhenMolecule implements moorhen.Molecule {
...
+        }
+        this.representations.forEach(representation => representation.drawSymmetry())
+    }
+
     /**
      * Draw symmetry mates for the current molecule
      * @param {boolean} [fetchSymMatrix=true] - Indicates whether a new symmetry matrix must be fetched from libcoot api
      */
     async drawSymmetry(fetchSymMatrix: boolean = true): Promise<void> {
+        if(this.biomolOn) return;
```
Comment (FilomenoSanchez) [inline baby-gru/src/utils/MoorhenMolecule.ts:294]: Can we print a warning? Something like `console.warn('Cannot draw cryst. symm. when bio. symm. is enabled)` will suffice.

File: baby-gru/src/utils/MoorhenMolecule.ts [hunk 3]
```diff
@@ -209,6 +222,42 @@ export class MoorhenMolecule implements moorhen.Molecule {
...
 
+    /**
+     * Fetch the matrices representing the biomolecule of the current model from gemmi
+     */
+    fetchBiomoleculeMatrix() : void {
+        if (!this.biomolOn) {
+            this.symmetryMatrices = []
+        } else {
+            this.symmetryMatrices = []
+            const assemblies = this.gemmiStructure.assemblies
```
Comment (FilomenoSanchez) [inline baby-gru/src/utils/MoorhenMolecule.ts:233]: I think `assemblies.delete()` needs to be called when we are done wiith this vector

File: baby-gru/src/utils/MoorhenMolecule.ts [hunk 4]
```diff
@@ -209,6 +222,42 @@ export class MoorhenMolecule implements moorhen.Molecule {
...
+            const n_assembly = assemblies.size()
+            if(n_assembly>0){
+                const assembly = assemblies.get(0)
+                const generators = assembly.generators
```
Comment (FilomenoSanchez) [inline baby-gru/src/utils/MoorhenMolecule.ts:237]: Same as above, `generators.delete()` needs to be called.

File: baby-gru/src/utils/MoorhenMolecule.ts [hunk 5]
```diff
@@ -209,6 +222,42 @@ export class MoorhenMolecule implements moorhen.Molecule {
...
+                const n_gen = generators.size()
+                if(n_gen>0){
+                    const gen = generators.get(0)
+                    const operators = gen.operators
```
Comment (FilomenoSanchez) [inline baby-gru/src/utils/MoorhenMolecule.ts:241]: We need to call `operators.delete()`

File: baby-gru/src/utils/MoorhenMolecule.ts [hunk 6]
```diff
@@ -192,6 +194,17 @@ export class MoorhenMolecule implements moorhen.Molecule {
         return Promise.reject(cootResponse.data.result.status)
     }
 
+    /**
+     * Turn on/off molecule symmetry
+     */
+    toggleBiomolecule() : void {
+        this.biomolOn = !this.biomolOn;
+        if(this.biomolOn)
+            this.symmetryOn = false;
```
Comment (FilomenoSanchez) [inline baby-gru/src/utils/MoorhenMolecule.ts:203]: If we don't want cryst and bio symm at the same time, do we also need to have the same check in `toggleSymmetry` but the other way around?

**Review by FilomenoSanchez — CHANGES_REQUESTED** [review]

File: baby-gru/src/components/card/MoorhenMoleculeRepresentationSettingsCard.tsx (hunk 1 above)
Comment (stuartjamesmcnicholas) [reply baby-gru/src/components/card/MoorhenMoleculeRepresentationSettingsCard.tsx]: Done in 98e02208e3be2dfe1892e1d9bf9ac019e067b9d4.

File: baby-gru/src/utils/MoorhenMolecule.ts (hunk 2 above)
Comment (stuartjamesmcnicholas) [reply baby-gru/src/utils/MoorhenMolecule.ts]: Done in 98e02208e3be2dfe1892e1d9bf9ac019e067b9d4.

File: baby-gru/src/utils/MoorhenMolecule.ts (hunk 3 above)
Comment (stuartjamesmcnicholas) [reply baby-gru/src/utils/MoorhenMolecule.ts:235]: Done in 98e02208e3be2dfe1892e1d9bf9ac019e067b9d4.

File: baby-gru/src/utils/MoorhenMolecule.ts (hunk 4 above)
Comment (stuartjamesmcnicholas) [reply baby-gru/src/utils/MoorhenMolecule.ts:239]: Done in 98e02208e3be2dfe1892e1d9bf9ac019e067b9d4.

File: baby-gru/src/utils/MoorhenMolecule.ts (hunk 5 above)
Comment (stuartjamesmcnicholas) [reply baby-gru/src/utils/MoorhenMolecule.ts:243]: Done in 98e02208e3be2dfe1892e1d9bf9ac019e067b9d4.

File: baby-gru/src/components/card/MoorhenMoleculeRepresentationSettingsCard.tsx [hunk 7]
```diff
@@ -212,6 +221,11 @@ export const MoorhenMoleculeRepresentationSettingsCard = (props: {
                         checked={symmetryOn}
                         onChange={() => { setSymmetryOn(!symmetryOn) }}
                         label="Show symmetry mates" />
+                    <Form.Check
+                        type="switch"
+                        checked={biomolOn}
+                        onChange={() => { setBiomolOn(!biomolOn) }}
```
[the remark this answers is not in the document]
Comment (stuartjamesmcnicholas) [reply baby-gru/src/components/card/MoorhenMoleculeRepresentationSettingsCard.tsx]: Done in 98e02208e3be2dfe1892e1d9bf9ac019e067b9d4.

File: baby-gru/src/utils/MoorhenMolecule.ts (hunk 6 above)
Comment (stuartjamesmcnicholas) [reply baby-gru/src/utils/MoorhenMolecule.ts:203]: Done in 98e02208e3be2dfe1892e1d9bf9ac019e067b9d4.

**Pull request merged**